lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp

```diff
@@ -0,0 +1,430 @@
+//===- LowerMemory.cpp - Lower Memories -------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// This file defines the LowerMemories pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/Dominance.h"
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/STLFunctionalExtras.h"
+#include "llvm/Support/Parallel.h"
+#include <optional>
+#include <set>
+
+using namespace circt;
+using namespace firrtl;
+
+static const char dutAnnoClass[] = "sifive.enterprise.firrtl.MarkDUTAnnotation";
+
+// Extract all the relevant attributes from the MemOp and return the FirMemory.
+FirMemory getSummary(MemOp op) {
+  size_t numReadPorts = 0;
+  size_t numWritePorts = 0;
+  size_t numReadWritePorts = 0;
+  llvm::SmallDenseMap<Value, unsigned> clockToLeader;
+  SmallVector<int32_t> writeClockIDs;
+
+  for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+    auto portKind = op.getPortKind(i);
+    if (portKind == MemOp::PortKind::Read)
+      ++numReadPorts;
+    else if (portKind == MemOp::PortKind::Write) {
+      for (auto *a : op.getResult(i).getUsers()) {
+        auto subfield = dyn_cast<SubfieldOp>(a);
+        if (!subfield || subfield.fieldIndex() != 2)
+          continue;
+        auto clockPort = a->getResult(0);
+        for (auto *b : clockPort.getUsers()) {
+          if (auto connect = dyn_cast<FConnectLike>(b)) {
+            if (connect.dest() == clockPort) {
+              auto result =
+                  clockToLeader.insert({connect.src(), numWritePorts});
+              if (result.second) {
+                writeClockIDs.push_back(numWritePorts);
+              } else {
+                writeClockIDs.push_back(result.first->second);
+              }
+            }
+          }
+        }
+        break;
+      }
+      ++numWritePorts;
+    } else
+      ++numReadWritePorts;
+  }
```

Comment:
This is a copy of `MemOp::getSummary()` with changes to how the memory name is calculated, it is now whatever the name of the original memory way.  I originally edited this function in place, but that requires my changes to `CreateSiFiveMetadata` and `LowerToHW` to be added at the same time as this one.  Eventually I think we can pull in the `struct FirMemory` structure into this file in its entirety.